File: flask-backend/models/habit.py
from datetime import datetime
from .db import db

class Habit(db.Model):
    __tablename__ = 'habits'
    
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(255), nullable=False)
    description = db.Column(db.Text, nullable=True)
    type = db.Column(db.String(50), default='daily')  # daily, weekly, monthly
    target = db.Column(db.Float, default=1.0)
    unit = db.Column(db.String(50), nullable=True)
    reminderTime = db.Column(db.Time, nullable=True)
    color = db.Column(db.String(20), default='#3f51b5')
    icon = db.Column(db.String(50), default='star')
    startDate = db.Column(db.Date, default=datetime.utcnow().date)
    endDate = db.Column(db.Date, nullable=True)
    active = db.Column(db.Boolean, default=True)
    userId = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    createdAt = db.Column(db.DateTime, default=datetime.utcnow)
    updatedAt = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    progress = db.relationship('HabitProgress', backref='habit', lazy=True, cascade="all, delete-orphan")
    
    def __init__(self, name, userId, **kwargs):
        self.name = name
        self.userId = userId
        for key, value in kwargs.items():
            setattr(self, key, value)
    
    def to_dict(self):
        """Convert the habit object to a dictionary"""
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'type': self.type,
            'target': self.target,
            'unit': self.unit,
            'reminderTime': self.reminderTime.isoformat() if self.reminderTime else None,
            'color': self.color,
            'icon': self.icon,
            'startDate': self.startDate.isoformat() if self.startDate else None,
            'endDate': self.endDate.isoformat() if self.endDate else None,
            'active': self.active,
            'userId': self.userId,
            'createdAt': self.createdAt.isoformat() if self.createdAt else None,
            'updatedAt': self.updatedAt.isoformat() if self.updatedAt else None
        } 

File: server/models/Habit.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/db');
const User = require('./User');

const Habit = sequelize.define('Habit', {
  name: {
    type: DataTypes.STRING,
    allowNull: false,
    validate: {
      notEmpty: { msg: 'Please add a habit name' },
      len: { args: [1, 100], msg: 'Name cannot be more than 100 characters' }
    }
  },
  description: {
    type: DataTypes.STRING(500),
    allowNull: true
  },
  type: {
    type: DataTypes.ENUM('daily', 'weekly', 'monthly'),
    defaultValue: 'daily'
  },
  target: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 1,
    validate: {
      min: { args: [1], msg: 'Target must be at least 1' }
    }
  },
  unit: {
    type: DataTypes.STRING,
    defaultValue: 'times'
  },
  reminderTime: {
    type: DataTypes.STRING,
    allowNull: true
  },
  color: {
    type: DataTypes.STRING,
    defaultValue: '#4285F4'
  },
  icon: {
    type: DataTypes.STRING,
    defaultValue: 'repeat'
  },
  active: {
    type: DataTypes.BOOLEAN,
    defaultValue: true
  },
  startDate: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW
  },
  endDate: {
    type: DataTypes.DATE,
    allowNull: true
  }
}, {
  timestamps: true,
  createdAt: 'createdAt',
  indexes: [
    {
      fields: ['userId', 'active']
    }
  ]
});

const HabitProgress = sequelize.define('HabitProgress', {
  date: {
    type: DataTypes.DATEONLY,
    allowNull: false
  },
  value: {
    type: DataTypes.FLOAT,
    defaultValue: 0
  },
  notes: {
    type: DataTypes.STRING,
    allowNull: true
  },
  completed: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  }
}, {
  timestamps: true,
  createdAt: 'createdAt',
  indexes: [
    {
      fields: ['habitId', 'date']
    },
    {
      fields: ['userId', 'date']
    }
  ]
});

// Define associations
Habit.belongsTo(User, {
  foreignKey: {
    name: 'userId',
    allowNull: false
  },
  onDelete: 'CASCADE'
});

User.hasMany(Habit, {
  foreignKey: 'userId'
});

HabitProgress.belongsTo(Habit, {
  foreignKey: {
    name: 'habitId',
    allowNull: false
  },
  onDelete: 'CASCADE'
});

HabitProgress.belongsTo(User, {
  foreignKey: {
    name: 'userId',
    allowNull: false
  },
  onDelete: 'CASCADE'
});

Habit.hasMany(HabitProgress, {
  foreignKey: 'habitId'
});

User.hasMany(HabitProgress, {
  foreignKey: 'userId'
});

module.exports = { Habit, HabitProgress }; 